### PR TITLE
Fixed attributes keys cannot be accessed in Twig templates.

### PR DESCRIPTION
**Pull request by fabianmarz**
### Problem
- Attribute keys were converted in the parent template already which prevent direct key access

### Goal
- Make attribute keys direct accessible to check if attributes exist

### Proposed solution
- Do not convert attribute keys in parent templates to maintain direct key access

### CODE REVIEW

File: src/attributes.js [hunk 1]
```diff
@@ -137,41 +137,42 @@ class Attributes {
...
-    static convert(context) {
+    static convert(context, recurse) {
+        recurse = recurse || false;
         _.forEach(context, (value, name) => {
             if (typeof name === 'string' && name.indexOf('attributes') > -1) {
                 context[name] = new Attributes(value);
             }
-            else if (_.isObject(value)) {
+            else if (recurse && _.isObject(value)) {
                 this.convert(value);
```
Comment (sun) [inline src/attributes.js:146]:
```suggestion
                this.convert(value, recurse);
```

**Review by sun — COMMENTED** [review]
I would rather implement the Proxy class and always ensure that the keys are at least accessible.

**Comment by fabianmarz** [comment]
@sun, I reverted the change as the implementation seemed correct. As the attribute key was missing in the `.yml` config, the attribute object was not instanced from the Attribute class. This PR now still contains some code restructuring which can be committed.

**Comment by sun** [comment]
@fabianmarz Should we move upgrading of variables named "*attributes" to a different place in the processing, so that it does neither rely on the default context variables nor on the passed variables but instead on the final/resulting template variables?

**Comment by fabianmarz** [comment]
@sun, I added a JS proxy to get proper access to the keys. Please have a look again. As this is fairly attribute related I decided to have another method in the respective `Attributes` class.

**Comment by sun** [comment]
> I added a JS proxy to get proper access to the keys

Not sure I understand; does this fix the attributes access issue without adding the attributes subkey to the config.yml files?

**Comment by fabianmarz** [comment]
@sun, this will achieve the keys of the attribute class are accessible directly without having to write `content.attributes.storage.src` similar to the Drupal functionality

```twig
{{ dump(content.attributes)|raw }}
{{ dump(content.attributes.src)|raw }}
```

```
object(2) { [classes]=> object(0) { } [storage]=> object(1) { [src]=> string(34) "http://lorempixel.com/640/480/cats" } }
string(34) "http://lorempixel.com/640/480/cats"
```

**Comment by fabianmarz** [comment]
I encountered a `Maximum call stack size exceeded` error. Please don't review.

**Comment by fabianmarz** [comment]
@sun, ready to review 👍

File: src/attributes.js [hunk 2]
```diff
@@ -140,38 +140,72 @@ class Attributes {
...
+                      // If the classes are requested, don't access the storage property.
+                      if (name === 'class') {
+                          name = 'classes';
+                      }
+                      // Return key from storage property otherwise.
+                      else {
+                          target = target.storage;
+                      }
+                  }
+                  // Fail if property not exists on target.
```
Comment (sun) [inline src/attributes.js]:
```suggestion
                  // Do not forward other property accesses and tests (like isset()).
```

File: src/tags/index.js [hunk 3]
```diff
@@ -91,6 +91,7 @@ module.exports = function(fractal){
                         }
                     }
 
+                    // @todo prevent mutliple convert and merge attributes from Twig render method only.
```
Comment (sun) [inline src/tags/index.js]:
```suggestion
```

File: src/attributes.js [hunk 4]
```diff
@@ -140,38 +140,72 @@ class Attributes {
...
+  static proxy(attributes) {
+      return new Proxy(attributes, {
+          get (target, name, receiver) {
+              if (!Reflect.has(target, name)) {
+                  // Check if getter is string.
+                  if (typeof name === 'string' && name.indexOf('get') !== -1) {
+                      // Property names look like this:`isSrc`, `getSrc` so we need
+                      // to strip the get prefix to obtain the correct attribute key.
+                      name = name.replace('get', '').toLowerCase();
+                      // If the classes are requested, don't access the storage property.
```
Comment (sun) [inline src/attributes.js]:
```suggestion
                      // Re-route into storage unless class property is requested.
```

File: src/attributes.js [hunk 5]
```diff
@@ -140,38 +140,72 @@ class Attributes {
...
+                      if (name === 'class') {
+                          name = 'classes';
+                      }
+                      // Return key from storage property otherwise.
```
Comment (sun) [inline src/attributes.js]:
```suggestion
```

File: src/attributes.js [hunk 6]
```diff
@@ -140,38 +140,72 @@ class Attributes {
...
+  /**
+   * Proxies direct attribute property access.
+   *
+   * @param attributes
+   * @returns {any|undefined}
+   */
+  static proxy(attributes) {
+      return new Proxy(attributes, {
+          get (target, name, receiver) {
+              if (!Reflect.has(target, name)) {
```
Comment (sun) [inline src/attributes.js]:
```suggestion
              if (typeof name === 'string' && !Reflect.has(target, name)) {
```

File: src/attributes.js [hunk 7]
```diff
@@ -140,38 +140,72 @@ class Attributes {
...
+                  // Check if getter is string.
+                  if (typeof name === 'string' && name.indexOf('get') !== -1) {
```
Comment (sun) [inline src/attributes.js]:
```suggestion
                  if (name.indexOf('get') !== -1) {
```

File: src/attributes.js [hunk 8]
```diff
@@ -140,38 +140,72 @@ class Attributes {
...
+   * Proxies direct attribute property access.
+   *
+   * @param attributes
+   * @returns {any|undefined}
+   */
+  static proxy(attributes) {
+      return new Proxy(attributes, {
+          get (target, name, receiver) {
+              if (!Reflect.has(target, name)) {
+                  // Check if getter is string.
```
Comment (sun) [inline src/attributes.js]:
```suggestion
```

**Review by sun — APPROVED** [review]
@fabianmarz The new code was wrongly indented. I fixed that and also improved the signal/noise ratio of inline comments.

File: src/attributes.js [hunk 9]
```diff
@@ -181,22 +181,20 @@ class Attributes {
...
                       if (name === 'class') {
                           name = 'classes';
                       }
-                      // Return key from storage property otherwise.
                       else {
                           target = target.storage;
                       }
                   }
-                  // Fail if property not exists on target.
+                  // Do not forward other property accesses and tests (like isset()).
```
Comment (fabianmarz) [inline src/attributes.js]: @sun, `isset()` is not JS function, but this project is related to JS only.

File: src/tags/index.js [hunk 10]
```diff
@@ -91,7 +91,6 @@ module.exports = function(fractal){
                         }
                     }
 
-                    // @todo prevent mutliple convert and merge attributes from Twig render method only.
```
Comment (fabianmarz) [inline src/tags/index.js]: Why did you remove this comment?

Comment (sun) [reply src/tags/index.js]: Major refactoring tasks like this shouldn't be put in code as todos, but recorded as an issue instead; either here in this repo or elsewhere in the backlog of a project that needs it.
